woocommerce/woocommerce
# Product backorder stock status

Closes #11259 

TODO
<details>

- [x] Register new stock status named on-backorder
- [x] Update data stores to accept backorder status
- [x] Update `validate_props` for products to handle backorder status
- [x] Update variable data store sync methods
- [x] Update product admin displays to use the new status
- [x] Update product frontend display to use new status
- [x] Set on-backorder status based on stock and backorder setting criteria as outlined in #11259 
- [x] Allow filtering by stock status in admin menu
- [x] Hide manual stock status options and go full auto if managing stock.
- [x] Update routines to set correct status after the update

</details>

- [x] Update logic in the API classes (with backwards compatibility handling if needed) (Punted to API v3)
- [x] Update logic in the CSV importer/exporter (with backwards compatibility handling if needed)

I'll do these last couple TODO items in a separate PR once this is merged.

If this looks good, don't merge it yet. I'll update all of the phpcs stuff in these files and then we can merge. I didn't want my changes to get lost among all of the stuff we have to change to make phpcs stop complaining.

To test:
1. Bust CSS + JS caches.
2. Enable stock management for a product. 
3. Notice that you can't manually control the stock status any more. 
4. Set the stock quantity to 0 or less. 
5. Pick a different product. Disable stock management for that product.
6. Set the stock status for the product to On Backorder.
7. Go to the Edit Products page.
8. Notice that the stock status shows as On Backorder for the two products.
9. Filter by stock status using the new filter option on the Edit Products page.
